# Solution: two tree nodes on a file you have not seen

A second sample of Kalpa Retail orders arrived after the session. It is the same shape as the one
the class used and none of the numbers are the same, so nothing from the demo can be pasted in.

Every blank is filled and the notebook is executed, so the numbers below are the ones the file
actually holds. Read the reasons, not just the values.
were right, so you know before anybody marks it.

**Post the four letters** for the multiple-choice section at the end, in order, as one string.

In [1]:
import pathlib
import sys

root = next(p for p in pathlib.Path.cwd().resolve().parents if (p / "scripts" / "c2kit.py").exists())
sys.path.insert(0, str(root / "scripts"))
import c2kit as kit

ORDERS = kit.load_records("C2_W01_D01_takehome_STUDENT.py")
print(f"{len(ORDERS)} orders loaded")

24 orders loaded


## 1. Draw the tree before you compute anything

Fill the five branch labels. The edge label is what this file can say about that branch, or the
word `unknown` when the file cannot say anything.

In [2]:
kit.tree(
    {"label": "REVENUE",
     "branches": [
         ("19", {"label": "customers"}),
         ("1.26", {"label": "orders per customer"}),
         ("unknown", {"label": "items per order"}),
         ("unknown", {"label": "price per item"}),
         ("unknown", {"label": "discounts"}),
     ]},
    title="Your tree, before any code runs")

## 1b. Which "sales" are you about to compute?

Four honest answers live inside the word. Put them in order from the largest number to the
smallest, and light the one you are going to report.

In [3]:
kit.ladder(["gross bookings, everything ordered",
            "delivered revenue, what Finance recognises",
            "net of returns",
            "cash collected, which lags delivery"], lit=1,
           title="Four readings of the word sales, largest first")

## 2. Revenue

One amount in this file may not be a number. Decide what to do about it before you write the
loop, and write the reason in the comment.

In [4]:
# Reason for how I handle a non-numeric amount: one row stores its amount as text, so int() is
# applied to every amount rather than to that one row, because a rule that fires on one row today
# will miss the second row tomorrow. The decision is written here because on Wednesday it gets a log.
revenue = 0
for order in ORDERS:
    revenue = revenue + int(order["amount"])
print(f"revenue: Rs {revenue:,}")

revenue: Rs 569,540


In [5]:
kit.check("revenue is a whole number of rupees", isinstance(revenue, int), f"got {type(revenue).__name__}")
kit.check("revenue is above Rs 5 lakh", revenue > 500000, f"got Rs {revenue:,}")

## 3. Customers, and the rate that needs its denominator

In [6]:
customers = set()
for order in ORDERS:
    customers.add(order["customer_id"])

orders_per_customer = len(ORDERS) / len(customers)
print(f"customers: {len(customers)}")
print(f"orders per customer: {orders_per_customer:.2f}")

customers: 19
orders per customer: 1.26


In [7]:
kit.check("every customer counted once", len(customers) < len(ORDERS),
          f"{len(customers)} customers against {len(ORDERS)} orders")
kit.check("orders per customer is at least 1", orders_per_customer >= 1,
          f"got {orders_per_customer:.2f}")

## 4. The median, by hand

Sort the amounts, then take the middle one. With an even number of orders, take the average of
the two in the middle.

In [8]:
amounts = sorted(int(o["amount"]) for o in ORDERS)
middle = len(amounts) // 2
median = amounts[middle] if len(amounts) % 2 else (amounts[middle - 1] + amounts[middle]) / 2
mean = sum(amounts) / len(amounts)
print(f"mean:   Rs {mean:,.0f}")
print(f"median: Rs {median:,.0f}")

mean:   Rs 23,731
median: Rs 2,765


In [9]:
kit.check("the median sits inside the ordinary range", 500 < median < 6000, f"Rs {median:,.0f}")
kit.check("the mean is well above the median", mean > 4 * median,
          f"mean Rs {mean:,.0f} against median Rs {median:,.0f}")

## 5. Show the gap

Draw the four steps that explain why the mean and the median disagree on this file.

In [10]:
kit.flow(["24 orders", "sort by amount",
          "the top order is 55 percent", "report the median"], lit=2,
         title="Why the mean and the median disagree here")

## 6. The four letters to post

Answer each with one letter and post the four as one string, for example `BCAD`.

**Q1.** This file's mean order value is well above its median. What does that tell you?
`A` most orders are large  ·  `B` a few orders are far larger than the rest  ·
`C` the data is wrong  ·  `D` the median was computed incorrectly

**Q2.** You report one number to describe a typical order. Which?
`A` the mean  ·  `B` the median  ·  `C` the largest  ·  `D` the total divided by customers

**Q3.** `orders per customer` needs which denominator?
`A` all orders  ·  `B` delivered orders  ·  `C` distinct customers in the same window  ·
`D` all customers who ever bought

**Q4.** Which branch of the tree can this file say nothing about?
`A` customers  ·  `B` orders per customer  ·  `C` items per order  ·  `D` revenue

In [11]:
my_answers = "BBCC"
kit.check("four letters posted", len(my_answers) == 4 and my_answers.isalpha(),
          f"got {my_answers!r}")
kit.check_summary()

## 7. One sentence

Write the sentence you would send Meera about this sample: the claim, the caveat that would
change it, and what you would look at next. Under forty words.

> "On this sample the typical order is about Rs 2,800 and one corporate order carries 55 percent
> of the revenue, so the average of Rs 23,731 describes no order in the file. The caveat is that
> twenty-four orders is too few to say anything about a trend. I would next ask for a second
> period so we can see which branch moved."

## Why each letter

| | The key | Why the others fail |
|---|---|---|
| **Q1** | `B` | A says most orders are large, which the median disproves. C blames the data when the data is fine. D blames the arithmetic when the arithmetic is right. |
| **Q2** | `B` | The mean is dragged by one corporate order. The largest describes one row. Total over customers answers a different question, revenue per customer. |
| **Q3** | `C` | All orders is the numerator, not the denominator. Delivered orders changes what you are counting. Everyone who ever bought spans a different window, so the rate is not comparable. |
| **Q4** | `C` | Customers and orders per customer are computable here. Revenue is computable. Items per order needs an items table this file does not carry. |